vumi_twilio_api/twilio_api.py

```diff
@@ -66,6 +88,103 @@ def teardown_application(self):
         yield self.webserver.loseConnection()
         yield self.session_manager.stop()
 
+    def _http_request(self, url='', method='GET', data={}):
+        return treq.request(method, url, persistent=False, data=data)
+
+    def _request_data_from_session(self, session):
+        return {
+            'CallSid': session['CallId'],
+            'AccountSid': session['AccountSid'],
+            'From': session['From'],
+            'To': session['To'],
+            'CallStatus': session['Status'],
+            'ApiVersion': self.config.api_version,
+            'Direction': session['Direction'],
+        }
+
+    @inlineCallbacks
+    def _get_twiml_from_client(self, session):
+        data = self._request_data_from_session(session)
+        twiml_raw = yield self._http_request(
+            session['Url'], session['Method'], data)
+        if twiml_raw.code < 200 or twiml_raw.code >= 300:
+            twiml_raw = yield self._http_request(
+                session['FallbackUrl'], session['FallbackMethod'], data)
+        twiml_raw = yield twiml_raw.content()
+        returnValue(self.twiml_parser.parse(twiml_raw))
+
+    @inlineCallbacks
+    def _handle_connected_call(
+            self, session_id, session, status='in-progress'):
+        # TODO: Support sending ForwardedFrom parameter
+        # TODO: Support sending CallerName parameter
+        # TODO: Support sending geographic data parameters
+        session['Status'] = status
+        self.session_manager.save_session(session_id, session)
+        twiml = yield self._get_twiml_from_client(session)
+        for verb in twiml:
+            self._handle_twiml_verb(verb)
+
+    def _handle_twiml_verb(self, verb):
+        pass
+
+    @inlineCallbacks
+    def consume_ack(self, event):
+        message_id = event['user_message_id']
+        message = yield self.message_store.get_outbound_message(message_id)
+        session = yield self.session_manager.load_session(message['to_addr'])
+
+        if session['Status'] == 'queued':
+            yield self._handle_connected_call(message['to_addr'], session)
+
+    @inlineCallbacks
+    def consume_nack(self, event):
+        message_id = event['user_message_id']
+        message = yield self.message_store.get_outbound_message(message_id)
```

Comment:
The message store and Riak are really heavy dependencies and really very much overkill for what we need. Instead I think we should store a Redis key mapping the message_id to the session_id. We can set it to expire after some time. More long term once Vumi messages have a session id, we could perhaps consider allowing events to have session ids too.